src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,681 @@
+	codec_context->height        = height;
+	codec_context->time_base.num = 1;
+	codec_context->time_base.den = frames_per_second;
+	codec_context->framerate.num = frames_per_second;
+	codec_context->framerate.den = 1;
+	codec_context->pix_fmt       = AV_PIX_FMT_YUV420P;
+
+	if (avcodec_open2(codec_context, codec, nullptr) < 0) {
+		LOG_ERR("FFmpeg: Failed to open video context");
+		return false;
+	}
+
+	frame = av_frame_alloc();
+	if (!frame) {
+		LOG_ERR("FFmpeg: Failed to allocate video frame");
+		return false;
+	}
+	frame->width  = width;
+	frame->height = height;
+	frame->format = codec_context->pix_fmt;
+	frame->pts    = 0;
+	if (av_frame_get_buffer(frame, 0) < 0) {
+		LOG_ERR("FFmpeg: Failed to get video frame buffer");
+		return false;
+	}
+
+	// Initalise the re-scaler. Currently only used to convert to YuV color space.
+	rescaler_contex = sws_getContext(width,
+	                                 height,
+	                                 map_pixel_format(pixel_format),
+	                                 width,
+	                                 height,
+	                                 AV_PIX_FMT_YUV420P,
+	                                 SWS_BILINEAR,
+	                                 nullptr,
+	                                 nullptr,
+	                                 nullptr);
+	if (!rescaler_contex) {
+		LOG_ERR("FFmpeg: Failed to get swscaler context");
+		return false;
+	}
+
+	this->width             = width;
+	this->height            = height;
+	this->pixel_format      = pixel_format;
+	this->frames_per_second = frames_per_second;
+	return true;
+}
+
+void FFmpegVideoEncoder::free()
+{
+	if (rescaler_contex) {
+		sws_freeContext(rescaler_contex);
+		rescaler_contex = nullptr;
+	}
+	if (frame) {
+		av_frame_free(&frame);
+		frame = nullptr;
+	}
+	if (codec_context) {
+		avcodec_free_context(&codec_context);
+		codec_context = nullptr;
+	}
+}
+
+static void send_packets_to_muxer(AVCodecContext* context, int stream_index,
+                                  RWQueue<AVPacket*>& queue)
+{
+	int packet_recieved;
+	do {
+		AVPacket* packet = av_packet_alloc();
+		if (!packet) {
+			LOG_ERR("FFmpeg: Failed to allocate audio packet");
+			return;
+		}
+		packet_recieved = avcodec_receive_packet(context, packet);
+		if (packet_recieved == 0) {
+			packet->stream_index = stream_index;
+			if (!queue.Enqueue(std::move(packet))) {
+				// Should not ever happen
+				LOG_ERR("FFmpeg: Muxer queue is stopped");
+				av_packet_free(&packet);
+			}
+		} else {
+			av_packet_free(&packet);
+		}
+	} while (packet_recieved == 0);
+	// These are two legitamite errors that don't need to be logged.
+	// EAGAIN meaning there's no more packets for us right now.
+	// AVERROR_EOF happens once when we flush the encoders.
+	if (packet_recieved != AVERROR(EAGAIN) && packet_recieved != AVERROR_EOF) {
+		LOG_ERR("FFmpeg: Receive packet failed");
+	}
+}
+
+void FFmpegEncoder::encode_video()
+{
+	for (;;) {
+		std::unique_lock<std::mutex> lock(mutex);
```

Comment:
Is it intentional that we acquire the lock inside the for loop?

Reply:
That is intentional.  Once it finishes encoding an entire file, it takes the lock to wait until the main thread re-initializes the encoder with another file.

Reply:
Yeah I realised later; this is the worker function. The method name threw me off a bit 😄